Lantern android

xiam opened the pull request:
The [lantern-android](https://github.com/getlantern/lantern-android) project provides tools for creating a flashlight library that makes easy running and controlling a flashlight client on Android devices.

This commit provides example code and instructions.

oxtoacart commented:
@xiam I'm starting to review this now.  I may not be able to finish before I leave on vacation, but I'll start taking notes.

This is an awesome PR!  The README is excellent.

oxtoacart commented on src/github.com/getlantern/lantern-android/README.md:
```diff
@@ -0,0 +1,232 @@
+
+Now add a new *main activity* by right-clicking on the top most directory on
+the *Project* pane and selecting New->Activity->Blank Activity, the default
+values would be OK, click *Finish*.
```
@xiam This option is grayed out for me and Android studio complains that I need to set the minSdkVersion to at least 7.

If I update the defaultconfig in app/build.gradle to look like the below, then I can create an Activity:

``` groovy
    defaultConfig {
        applicationId "org.getlantern.example"
        minSdkVersion 7
    }
```

xiam replied:
@oxtoacart I'm not sure what's causing that but maybe I can find a fix if I could replicate it, what's your Android Studio version? mine is this:

![untitled](https://cloud.githubusercontent.com/assets/385670/5929571/7c5c8926-a64a-11e4-93c6-41b7e9599dfc.png)

oxtoacart replied:
Weird, mine is older than yours, even though I just updated it ...

![screen shot 2015-01-27 at 5 54 38 pm](https://cloud.githubusercontent.com/assets/5000654/5929831/bb608052-a64d-11e4-9370-32cf063e4587.png)

oxtoacart replied:
Hmm, I just checked again and an update is available.  I installed it, and now I'm at this:

![screen shot 2015-01-27 at 5 56 58 pm](https://cloud.githubusercontent.com/assets/5000654/5929842/e95d4b2a-a64d-11e4-8262-bb6e53a058fb.png)

xiam commented:
@oxtoacart Thanks! actually the README on the main repo was a bit outdated compared to the copy I prepared for this PR but I just updated it :-).

oxtoacart commented on src/github.com/getlantern/lantern-android/README.md:
```diff
@@ -0,0 +1,232 @@
+
+The `make` command will create a new `app` subdirectory that will contain an
+Android example project. You may import the contents of the `app` subdirectory
+into Android Studio to see libflashlight working.
```
@xiam It turns out that I can't run the Android emulator on my Mac if VirtualBox is open.  I forget whether I opened VirtualBox myself or if boot2docker did it during the make process, but it might be nice to add a note about that in here.

xiam replied:
@oxtoacart  You're right, I'll also need to add a note specifying that the ARM emulator is required, the default is the x86 emulator.

xiam replied:
I could not reproduce this error:

![screen shot 2015-01-30 at 6 31 42 pm](https://cloud.githubusercontent.com/assets/385670/5985846/4af7d212-a8ae-11e4-976c-a4db92b1fa39.png)

does this happen every time you try to run the Android emulator when VirtualBox is already opened?

oxtoacart commented on src/github.com/getlantern/lantern-android/README.md:
````diff
@@ -0,0 +1,232 @@
+```
+curl -x 10.10.100.97:9192 http://www.google.com/humans.txt
+# Google is built by a large team of engineers, designers, researchers, robots, and others in many different sites across the globe. It is updated continuously, and built with more tools and technologies than we can shake a stick at. If you'd like to help us out, see google.com/careers.
+```
````
I'm not able to make this work, as I don't have access to the emulator's network.

What I can do however is set up redirection by first telnetting to the emulator's console and then setting up a redirect, like this:

``` bash
➜  ox.to.a.cart git:(master) ✗ telnet localhost 5556
Trying ::1...
telnet: connect to address ::1: Connection refused
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
Android Console: type 'help' for a list of commands
OK
redir add tcp:9192:9192
OK
```

Then I can

```
curl -x localhost:9192 http://www.google.com/humans.txt
```

oxtoacart commented on src/github.com/getlantern/lantern-android/client/client.go:
```diff
@@ -0,0 +1,209 @@
+// AddFrontedServer adds a fronted server to the list.
+func (client *Client) AddFrontedServer(fs *frontedServer) error {
+	client.frontedServers = append(client.frontedServers, fs)
+	// TODO: Find the best way to add this server to the balancer list.
```
@xiam Balancer was meant to be initialized through an idempotent operation (i.e. supplying a full list of dialers).  If the list ever changes, you can just stop the old balancer and initialize a new balancer using the new full list.  I've deliberately followed this pattern in a variety of places to reduce the complexity of individual adds and removes cascading down through the layers.

xiam replied:
Thanks for explaining @oxtoacart, then we may not need this. I think we're going to use one client on Android so I'll drop this.

oxtoacart commented on src/github.com/getlantern/lantern-android/client/client.go:
```diff
@@ -0,0 +1,209 @@
+}
+
+// AddFrontedServer adds a fronted server to the list.
+func (client *Client) AddFrontedServer(fs *frontedServer) error {
```
Why is this exported?  Is this actually a public API?

xiam replied:
The `Client` type must be exported so it can be used in the library bindings: https://github.com/getlantern/lantern-android/blob/master/libflashlight/bindings/main.go, but I see your point and certainly other types this package is exporting such as `CA`, `Config` and `Listener` aren't mean to be used outside the client so I'd better fix them.

oxtoacart replied:
Makes sense. My question was actually about the `AddFrontedServer` method specifically.  Is that something that's part of the public API here?

xiam replied:
No, it's not. I made it public before actually using the config file, but it really does not make sense now so I'll drop to too.

oxtoacart commented on src/github.com/getlantern/lantern-android/client/client_test.go:
```diff
@@ -0,0 +1,108 @@
+var testURLs = map[string][]byte{
+	"http://www.google.com/humans.txt": []byte("Google is built by a large team of engineers, designers, researchers, robots, and others in many different sites across the globe. It is updated continuously, and built with more tools and technologies than we can shake a stick at. If you'd like to help us out, see google.com/careers.\n"),
+	// TODO: This is not working, we actually need to implement a CONNECT request to proxy HTTPs traffic.
+	// "https://www.google.com/humans.txt": []byte("Google is built by a large team of engineers, designers, researchers, robots, and others in many different sites across the globe. It is updated continuously, and built with more tools and technologies than we can shake a stick at. If you'd like to help us out, see google.com/careers.\n"),
```
@xiam You should be able to configure the http.Transport with a proxy, and it will take care of doing the CONNECT request for you.

xiam replied:
I initially though this configuration was correct, but I can't make it work for HTTPs proxying.

``` go
        Transport: &http.Transport{
            Dial: func(n, a string) (net.Conn, error) {
                return net.Dial("tcp", listenProxyAddr)
            },
        },
```

Can you spot what I'm missing or point me to another similar test I could read?

Thanks!

atavism replied:
Hi @xiam, since @oxtoacart is on vacation, I'll chime in, but he can correct me if I happen to be mistaken. You'll want to configure http.Transport with a Proxy: 

``` go
    client := &http.Client{
        Transport: &http.Transport{
            Proxy: func(req *http.Request) (*url.URL, error) {
                return url.Parse(listenProxyAddr)
            },
            Dial: func(n, a string) (net.Conn, error) {
                return net.Dial("tcp", listenProxyAddr)
            },
        },
    }
```

A similar example can be found in flashlight_test.go

xiam replied:
Cool! now I understand, thank you very much @atavism!

oxtoacart commented on src/github.com/getlantern/lantern-android/client/client_test.go:
```diff
@@ -0,0 +1,108 @@
+	return nil
+}
+
+func TestListenAndServeStop(t *testing.T) {
```
What is this testing?  It doesn't have any assertions.

xiam replied:
I was used to create tests relying on the `testing` package only, do you think I could improve my tests by using `github.com/getlantern/testify/assert` and adding appropriate descriptions?

oxtoacart replied:
I find the stuff written using assert to be easier to read and debug, but I don't think it's required.

My question here was actually about the `TestListenAndServeStop` method.  It calls some APIs, but it doesn't actually check anything (e.g. return values, ability to connect/proxy via a port, file descriptor counts, etc.), so it's not clear to me that it's actually testing anything.

xiam replied:
After reading some of your tests on the code review I think I've learnt a better way to distribute my own tests, I'm sure this will be better after a bit of refactoring.

oxtoacart commented on src/github.com/getlantern/lantern-android/client/config.go:
```diff
@@ -0,0 +1,122 @@
+	var res *http.Response
+
+	// Issuing a post request to download configuration file.
+	if res, err = http.Get(remoteConfigURL); err != nil {
```
This will rarely work in the censored world.  We have to assume that the site from which we're fetching the updated configuration is blocked.  In fact, the site we use right now is s3.amazonaws.com, and that is periodically blocked in China.  That's why flashlight uses proxy for the http client - https://github.com/getlantern/flashlight-build/blob/peerscanner/src/github.com/getlantern/flashlight/config/config.go#L290

What flashlight actually does is the following:
1. Start using a list of servers and masquerades that's compiled into the binary (I see that you have that list, but aren't actually starting with it unless you're unable to fetch the list from the cloud)
2. Periodically fetch updates from the cloud by making proxied http requests (I see that you're only checking once, which is okay for now)
3. Update its configuration whenever a change has been detected in the cloud
4. Save the configuration to disk for whenever it next starts up (it looks like you're not saving it anywhere?)

The idea behind this is that the binary that we distribute will work for a while, but we may have to change our set of masquerades and servers to stay ahead of the censors.  As long as we're actually staying ahead of the censors, this will keep any Lantern clients that are already running unblocked.  And since these clients save their configuration to disk, they should remain unblocked even after a restart.

Ultimately, I think we'll want to do something similar for the Android version, but we don't need to be that sophisticated yet.  I think the way that you have should be fine with two changes:
1. Immediately start proxying based on the built-in configuration
2. Check for the cloud configuration through the already running proxy

xiam replied:
@oxtoacart Yeah, you're right, we can't suppose we're going to have this URL always available.

> Immediately start proxying based on the built-in configuration
> Check for the cloud configuration through the already running proxy

This sounds good, I could give it another try.

> Save the configuration to disk for whenever it next starts up (it looks like you're not saving it anywhere?)

We have a limitation on Android and is that writing files is not that easy unless we have a writable path, as far as I know, this path is different for each application namespace, and we don't have that information unless the app provides it to us. We could export a new function that the app could use to set this path and we could save the configuration file, but I'm not sure if this would be the best solution on the Android world.

oxtoacart replied:
> We have a limitation on Android and is that writing files is not that easy unless we have a writable path, as far as I know, this path is different for each application namespace, and we don't have that information unless the app provides it to us. We could export a new function that the app could use to set this path and we could save the configuration file, but I'm not sure if this would be the best solution on the Android world.

Yeah.  It seems that in Android, it might be better to have the app itself manage the configuration and just pass it in to the Go side (perhaps just as one big config object).  Anytime the app wants to change the configuration, it can just pass in the whole new configuration.  This would actually be very similar to how flashlight works - anytime that any part of the configuration is updated, the whole configuration is propagated through the system*.  Individual components are free to ignore it if their part of it hasn't changed.
- \- In reality, only portions of the configuration are propagated, based on what each individual component needs/has access to.

xiam replied:
@oxtoacart (I'm just replying to advance on this issue before I took another task, but this is not a top priority :-))

I just pushed some changes that will enable pulling the cloud configuration file over proxies periodically.

This is currently implemented like this:
- When the library starts, it immediately attempts to pull a fresh configuration directly, if something fails (timed out, access denied by censors, untrusted ssl connection, etc) it will continue with the default configuration.
- When a proxy starts (this will be requested by the Android app), it will fire a goroutine that will periodically check for a new configuration file, this pull will be through the client proxy itself. If any change is detected (based on the `Etag` and `If-None-Match` headers), the client will reload its configuration the same way the original flashlight client does.

oxtoacart commented on src/github.com/getlantern/lantern-android/client/conn.go:
```diff
@@ -0,0 +1,63 @@
+package client
```
Maybe this file should be called listener.go?

oxtoacart commented on src/github.com/getlantern/lantern-android/client/conn.go:
```diff
@@ -0,0 +1,63 @@
+			return nil, ErrClosed
+		default:
+			return li.TCPListener.Accept()
+		}
```
I don't think this will work quite as intended.  In particular, TCPListener.Accept() will block until a new connection is made.  If no new connection is made, the loop will never get a chance to check li.closed.

Is there a reason you don't just use a regular listener and call its Close() method when you want to stop listening?

xiam replied:
> Is there a reason you don't just use a regular listener and call its Close() method when you want to stop listening?

No, it was a stupid mistake. I was actually trying to create an error within the `http.Server.Serve()` method, specifically on this line: http://golang.org/src/net/http/server.go#L1728 that stopped the http server, so I thought that I could wrap `li.Accept()` to make it produce an error when I wanted. But `Close()` produced the error I needed already. I could swear I made a test before with a plain `net.Listener` that didn't work and kept blocking instead.

Sorry, I don't know what was I thinking but it had to be something rather dumb. I removed this thing completely from the source (even though it will remain in my memory for a while, haha).

Thank you very much for pointing this out @oxtoacart :-).

oxtoacart commented on src/github.com/getlantern/lantern-android/client/conn.go:
```diff
@@ -0,0 +1,63 @@
+
+// Listener is a wrapper around net.TCPListener that attempts to provide a
+// Stop() function.
+type Listener struct {
```
As a rule of thumb, I prefer to not export anything unless it needs to be used outside of its package.  Listener seems like it's completely internal to the client package.

oxtoacart commented:
@xiam Very nice!  Other than my inline comments, the only overarching thing that I see is that this ends up duplicating a fair amount of the code in flashlight.  I don't think we need to resolve this for the current PR, but I think it might make sense to create a flashlight lib that does all of the core stuff that's common between flashlight and lantern-mobile (e.g. configuration, bootstrapping fronted, etc.).  Doing that well will require some thought, so there's no need to hurry it.

xiam commented:
@oxtoacart Thanks for taking the time to let me some comments before your vacation! Allow me some time to clean mi pull request.

> but I think it might make sense to create a flashlight lib that does all of the core stuff

Yes, I ended up copying/duplicating lots of functionality to get through. Creating a flashlight library would be great! I think it is not that difficult but it won't be straightforward, we need to reduce dependencies and think of the limitations we have on Android too, so yes, we need to discuss it and give it some thought.

oxtoacart commented:
> Thanks for taking the time to let me some comments before your vacation! Allow me some time to clean mi pull request.

You're welcome.

oxtoacart committed:
linux_compile fixes

oxtoacart committed:
linux_compile fixes

oxtoacart committed:
linux_compile fixes

oxtoacart committed:
linux_compile fixes

Commit:
Moving logging to its own package, so we can add a Configure() func to it.

oxtoacart committed:
Uninstall improvements

Commit:
Configuring loggly proxy based on flashlight's settings.

Commit:
Squashed 'src/github.com/getlantern/go-loggly/' content from commit b13ac8b

git-subtree-dir: src/github.com/getlantern/go-loggly
git-subtree-split: b13ac8b01aea568a028afa3576a8ab6cc0491b43

Commit:
Merge commit '557bb6938b0317190bd6349281fd4155b7fdd6ea' as 'src/github.com/getlantern/go-loggly'

Commit:
Merge branch 'gost/issue-2232' into issue-2232

Commit:
Adding mutex to protect Configure and removing unused httpClient.

Commit:
No need to use RWMutex.

Commit:
Merge pull request #90 from getlantern/valencia-wininstall

Packaging windows as installer

Commit:
display current version of lantern in the UI

oxtoacart committed:
Fixed nil masquerades issue and problem with calculating wait time

oxtoacart committed:
Added space before parenthesis in version string

oxtoacart committed:
Merge pull request #94 from getlantern/valencia-version

Display current version of lantern in the UI

oxtoacart committed:
Improved validation in build scripts for windows code signing

oxtoacart committed:
converting pacUrl to wchar_t and fixing http:// prefix for pac url closes getlantern/lantern#2260

Commit:
Adding some clarifications.

Commit:
Moving underlying HTTP client inside loggly.Client struct.

oxtoacart committed:
Removed excess symbol from date format string

oxtoacart committed:
Always udpating dist with tagAndBuild

oxtoacart committed:
getlantern/lantern#2232 Code review fixes

oxtoacart committed:
Fixed pathing of systray.dll

oxtoacart committed:
Fixed pathing of systray.dll

oxtoacart committed:
Rebuilt DLL with static linking and also signed it

oxtoacart committed:
Merge branch 'valencia' into issue2260

Commit:
Merge pull request #95 from getlantern/issue2260

converting pacUrl to wchar_t and fixing http:// prefix for pac url close...

oxtoacart committed:
Dynamically picking masquerade on each dial of direct fronted client

oxtoacart committed:
Removed country tooltips, closes getlantern/lantern#2265

oxtoacart committed:
Moved goroutine into logging, fixed timezone reporting

oxtoacart committed:
Took out tooltip placement attribute

Commit:
Merge pull request #99 from getlantern/issue2265

Removed country tooltips, closes getlantern/lantern#2265

oxtoacart committed:
Merged valencia

oxtoacart committed:
Merged valencia

oxtoacart committed:
Merged issue-2232

Commit:
Merge branch 'integration/issue-2232-issue-2232-ox' into issue-2232

Commit:
Merge branch 'valencia' into issue-2232

oxtoacart committed:
Merge pull request #93 from getlantern/issue-2232

Issue 2232: Report Go errors to Loggly

fffw committed:
remove tray icon before quit app

aranhoide committed:
add loggly token to linux compile

aranhoide committed:
script to make ubuntu installers

fffw committed:
fix on linux

oxtoacart committed:
Signed systray dll

oxtoacart committed:
Updated Quit() comment

oxtoacart committed:
Merge pull request #100 from getlantern/issue2267

Remove icon from systray when app exits

aranhoide committed:
update README, make lintian a bit happier

aranhoide committed:
.gitignore vim swap files

aranhoide committed:
first shot at description

oxtoacart committed:
Updated README to reflect that 32bit works too

oxtoacart committed:
Merge pull request #101 from getlantern/ubuntu-installer

Ubuntu installer

oxtoacart committed:
Removed unnecessary Visual Studio files

oxtoacart committed:
Using bzip compression for Lantern.dmg

oxtoacart committed:
getlantern/lantern#2210 Peerscanner now maintains country-specific fallback rotations

oxtoacart committed:
Switched to using nl.fallbacks.getiantem.org and made peers a smaller percentage of rotation

Commit:
Merge pull request #102 from getlantern/issue2210

getlantern/lantern#2210 Peerscanner now maintains country-specific fallb...

oxtoacart committed:
Ran genconfig

fffw committed:
exit if pac tool not properly setted up

fffw committed:
make password prompt shorter to fit in one line

fffw committed:
revert to NonStopWriter

fffw committed:
add google domains to genconfig

fffw committed:
add simple script to genconfig

fffw committed:
genconfig with google domains

oxtoacart committed:
Added error handling to nonStopWriter

oxtoacart committed:
Merge pull request #103 from getlantern/issue2257

exit if pac tool not properly setted up

fffw committed:
Merge pull request #104 from getlantern/issue2270

revert to NonStopWriter closes getlantern/lantern#2270

oxtoacart committed:
Cleaned up exit logic in flashlight.go

fffw committed:
only save pac prompt icon on darwin

fffw committed:
Merge pull request #106 from getlantern/exitgracefully

Cleaned up exit logic in flashlight.go

fffw committed:
double wait time in TestWaitUntilNoneMatchTimeout

myleshorton committed:
Merge pull request #107 from getlantern/issue2251

prevent random failed test of fdcount

oxtoacart committed:
getlantern/lantern#2274 getlantern/lantern#2276 Cloud configuration fixes

Commit:
Merge pull request #108 from getlantern/yamlfix

getlantern/lantern#2274 getlantern/lantern#2276 Cloud configuration fixe...

myleshorton committed:
simple run script closes getlantern/lantern#2277

oxtoacart committed:
peerscanner logging updates

oxtoacart committed:
Merged devel to valencia

oxtoacart committed:
Not merging cloud configuration if it's unchanged

oxtoacart committed:
Fixed spelling mistake in fronted log

oxtoacart committed:
Including more ip prefixes as fallbacks, closes getlantern/lantern#1739

oxtoacart committed:
Clearing registration information irrespective of whether or not deregistering works

fffw committed:
Merge commit 'ebcbbe9910dce05b863bd335d465773cec97fe48' as 'src/github.com/getlantern/detour'

fffw committed:
Squashed 'src/github.com/getlantern/detour/' content from commit f75a9ea

git-subtree-dir: src/github.com/getlantern/detour
git-subtree-split: f75a9ea985f052f15ce4b86a1d9e7cf2cd014636

oxtoacart committed:
Merged valencia into issue2269 and regenerated configs

oxtoacart committed:
Merge pull request #105 from getlantern/issue2269

add google domains to genconfig

myleshorton committed:
Merge pull request #110 from getlantern/issue1739

Including more ip prefixes as fallbacks, closes getlantern/lantern#1739

fffw committed:
prototype to detour

oxtoacart committed:
Moved build and run onto one line

oxtoacart committed:
Merge pull request #109 from getlantern/issue2277

Just adds a simple run script closes getlantern/lantern#2277

oxtoacart committed:
Updated favicon to be 32 pixels, closes getlantern/lantern#2272

fffw committed:
detect timeout

fffw committed:
pass tests

oxtoacart committed:
Removed dogenzaka rotator

oxtoacart committed:
Merge commit 'e5e35ed57c9505225721a82d72e1c0b705df693e' as 'src/github.com/getlantern/rotator'

oxtoacart committed:
Squashed 'src/github.com/getlantern/rotator/' content from commit 97947be

git-subtree-dir: src/github.com/getlantern/rotator
git-subtree-split: 97947bef5b937df540b59a9b83d3ce6b9760a138

oxtoacart committed:
Fixed log rotation

fffw committed:
integrate into flashlight

oxtoacart committed:
Code review fixes

aranhoide committed:
Merge pull request #111 from getlantern/logfix

Fix log rotation (required forking rotator)

oxtoacart committed:
Removed peers

aranhoide committed:
Merge pull request #112 from getlantern/peerfix

Removed peers

oxtoacart committed:
Added back peers at weight 0

oxtoacart committed:
Removed peers from cloud.yaml

oxtoacart committed:
Merged

fffw committed:
just works

fffw committed:
Merge branch 'valencia' into detour

fffw committed:
clean up

fffw committed:
fix unchanged name

oxtoacart committed:
getlantern/lantern#2280 Not registering peers, registering fallbacks as peers instead

oxtoacart committed:
Clearing record after removing from rotation

fffw committed:
more reliable version

fffw committed:
âclean up again

fffw committed:
more bug fix

aranhoide committed:
Merge pull request #115 from getlantern/issue2280

getlantern/lantern#2280 Not registering peers, registering fallbacks as ...

fffw committed:
init and dump whitelist

fffw committed:
bug fix

oxtoacart committed:
getlantern/lantern#2285 Only logging geolookup error as error if we've failed a lot

oxtoacart committed:
getlantern/lantern#2285 Added additional logging when geolookup fails

oxtoacart committed:
getlantern/lantern#2284 Hiding hostname from logs

fffw committed:
bug fix

fffw committed:
add test

fffw committed:
refactor

fffw committed:
improve golog

Commit:
Merge pull request #116 from getlantern/issue2285

Better logging for geolookup errors

oxtoacart committed:
closes getlantern/lantern#2288 Added headless mode for lantern

fffw committed:
extract prefix and truncate messages to 100 chars

fffw committed:
add unit test

fffw committed:
extract only last 2 chunks to message

fffw committed:
change prefix to 'flashlight.logging'

fffw committed:
add preventive code in case no prefix added

fffw committed:
code review update

oxtoacart committed:
Merge pull request #119 from getlantern/issue2287

extract prefix and truncate messages to 100 chars

oxtoacart committed:
getlantern/lantern#2296 Ignoring extra flag from OS X when running from quarantined

fffw committed:
Merge pull request #120 from getlantern/issue2296

getlantern/lantern#2296 Ignoring extra flag from OS X when running from ...

oxtoacart committed:
Merge commit '717294d33c4f6ecab8d880452a86c6aea540cf62' into valencia

oxtoacart committed:
Squashed 'src/github.com/getlantern/enproxy/' changes from a8e5b54..8e48af0

8e48af0 Added ability to wait for condition in fdcount
6daa733 getlantern/lantern#1739 peerscanner now verifies host in response from proxies
0a16190 getlantern/lantern#1739 peerscanner now verifies host in response from proxies
e90ba2f Attempt to fix race condition with a lingering CLOSE_WAIT.
b20ee6b Dumping connection trace if file descriptor count in enproxy test fails
97fb38f Removed unnecessary setDeadline in reading response to request
1cb39cd Revised connection tracing output
d92fbd9 Smaller buffers and moved open tracking
12f6539 Another tweak to the closing sequence
453968f Closing initialRequestCh from conn_impl_requests instead of conn_impl_writes
f43bc16 Removed default NewRequest
adf80d8 Made Conn package private
2a11152 More docs on tracing connection state
fbac79b Updated travis builds to 1.4.1
e9abc68 Cleaned up Dial function
1d4074b Added info about TRACE_CONN_STATE to README
0cab6b9 Enproxy working pertty darned well
8416b51 Added connection state profiling, made progress on cleaning up close sequence
bac12c7 Closing connection on fail
f659a0b Cleaned up NewRequest default
3bfcb96 Maybe fixed test
0ddee4c Factored out idling logic and just using idletiming
b95ae01 Merge branch 'issue-2178' into peerscanner
c991bf8 getlantern/lantern#2178 Added doneWritingCh
16f1729 Merged from issue-2178
31c0166 getlantern/lantern#2178 Rationlized enproxy Close processing
8f547f3 getlantern/lantern#2178 Added test and fix for leaking connections
e6cbd10 Fixed connection leak in enproxy
00adef8 Checking file descriptors in unit test
61c2d98 Merge branch '2165' into peerscanner
6d425f0 Cleaned up dialing interface on enproxy
abf13e1 getlantern/lantern#2165 Interrupting read/write processing if a request fails
cdccfcd Merge pull request #20 from getlantern/2117
5ca72e5 Merge branch 'tcpproxying' into 2135
c9eb5b5 getlantern/lantern#2117 Shifted responsibility for closing connections from idletiming.Conn to client of idletiming.Conn
41ea470 Merge branch 'devel' into tcpproxying
037d7fb Merge branch 'master' into devel
bc28dc6 Synchronized access to connMap
b5ac271 getlantern/lantern#2121 Breaking out of loop when timeout occurs before deadline set for idle timeout
bd7d728 getlantern/lantern#2135 Switched to using net.SplitHostAndPort
cc59cb6 getlantern/lantern#2135 Switched to using net.SplitHostAndPort
6f07234 getlantern/lantern#2135 Switched to using net.SplitHostAndPort
ac6aa3e Merge branch 'tcpproxying' into 2122
afd20b5 getlantern/lantern#2122 Code Review Updates
326f3ba getlantern/lantern#2122 Added more information to stat callbacks in enproxy
94f86d1 getlantern/lantern#2083 Switched travis builds to use golang.org/x/tools/cmd/cover
6552358 getlantern/lantern#2083 Switched travis builds to Go 1.4

git-subtree-dir: src/github.com/getlantern/enproxy
git-subtree-split: 8e48af0387c19acb54176fa39f616d140b0dd79f

oxtoacart committed:
Squashed 'src/github.com/getlantern/golog/' changes from 29b4121..5245e76

5245e76 Cleaned up exit logic in flashlight.go
5e80d36 Logging fixes from code review
126a963 Merged golog
b270fb1 Updated travis builds to 1.4.1

git-subtree-dir: src/github.com/getlantern/golog
git-subtree-split: 5245e761fd1c8e9ccba87de802a0f666ed55980f

oxtoacart committed:
Merge commit 'b2a4138956bdf9b0c0066439c2424aae90fc4942' into valencia

oxtoacart committed:
Merge commit '5543717179b1e9f887b49d27ffb286d9ec23fdad' into valencia

oxtoacart committed:
Squashed 'src/github.com/davecgh/go-spew/' changes from 1aaf839..fc32781

fc32781 Add -u to the README installation instructions.
a4b53de Add support for sorting keys that are arrays.

git-subtree-dir: src/github.com/davecgh/go-spew
git-subtree-split: fc32781af5e85e548d3f1abaf0fa3dbe8a72495c

oxtoacart committed:
Merge commit 'ca21ed5d3486f59277e9b5b49a62d5f7f936e53e' into valencia

oxtoacart committed:
Squashed 'src/github.com/getlantern/idletiming/' changes from b614bc6..0e09328

0e09328 Updated travis builds to 1.4.1
d70a5cb Merge branch 'devel' into tcpproxying
88b2887 Merge branch 'devel' into 2127
a2c5557 getlantern/lantern#2117 Cleaned up onIdle/onClose terminology
cc82ee1 getlantern/lantern#2117 Comment update
cdde516 getlantern/lantern#2117 More close cleanup
13b9987 getlantern/lantern#2117 Ensuring that close callback is called when Close() was called
52b34a3 getlantern/lantern#2117 Fixed Close method, added test and comment
f1c23a9 Fixed nil pointer in idletiming.Listener close handling
42ceb81 getlantern/lantern#2117 Shifted responsibility for closing connections from idletiming.Conn to client of idletiming.Conn
04d460d getlantern/lantern#2117 idletiming.Conn now closes the underlying connection after idling, also cleaned up a few other items and the unit test
0078356 Merge branch 'master' into devel
5d73e72 getlantern/lantern#2121 Breaking out of loop when timeout occurs before deadline set for idle timeout
9c35001 getlantern/lantern#2121 Synchronizing access to read and write deadlines, fetching them only once at the beginning of our loops
01099ba getlantern/lantern#2121 Idletiming checks type of timeout error (read vs write)
8c5507f getlantern/lantern#2121 Reduced CPU usage in waitforserver by increasing sleep time on each loop
00f647d getlantern/lantern#2127 Fixed race condition in idletiming conn
36e922a getlantern/lantern#2083 Switched travis builds to use golang.org/x/tools/cmd/cover
0fb497e getlantern/lantern#2083 Switched travis builds to Go 1.4

git-subtree-dir: src/github.com/getlantern/idletiming
git-subtree-split: 0e09328c5a420afbe2cfdb94a107acdd3190a189

oxtoacart committed:
Squashed 'src/github.com/getlantern/fdcount/' changes from 8c6e363..be4f757

be4f757 Added ability to wait for condition in fdcount
6f2f28e Attempt to fix race condition with a lingering CLOSE_WAIT.
08ca558 Added retry of fdcount.Assert with exponential back-off
a4fecc9 Reversed file descriptor debug info in error message
d990511 Updated travis builds to 1.4.1
7be201f Removed log from test
89083e6 Merged from issue-2178
86ceaae Better file descriptor diff reporting
cb2168e Updated lsof output to show delta only
7c3d4ae Added LICENSES to those projects that needed them

git-subtree-dir: src/github.com/getlantern/fdcount
git-subtree-split: be4f75758b4d7042adc6949091e0f05f5eb411ea

oxtoacart committed:
Merged enproxy from upstream

oxtoacart committed:
Merge commit 'd0f1706474634d55e6708d9eb354910ead31303b' into pushupstream

oxtoacart committed:
Squashed 'src/github.com/getlantern/buuid/' changes from 248cea1..43f979e

43f979e Updated travis builds to 1.4.1
e842af7 Added LICENSES to those projects that needed them
b562fe4 getlantern/lantern#2083 Switched travis builds to use golang.org/x/tools/cmd/cover
488d35e getlantern/lantern#2083 Switched travis builds to Go 1.4

git-subtree-dir: src/github.com/getlantern/buuid
git-subtree-split: 43f979e2596f6c91a759e25d3901e91fbde8bd96

oxtoacart committed:
[gost] Added github.com/getlantern/buuid and its dependencies

oxtoacart committed:
Squashed 'src/github.com/getlantern/bytecounting/' changes from 49e163a..a672d58

a672d58 Updated travis builds to 1.4.1
1150d87 Switched to golang.org/x/tools/cmd/cover
dd86aa9 Upgraded remaining projects to Go 1.4

git-subtree-dir: src/github.com/getlantern/bytecounting
git-subtree-split: a672d582e31fcd119bd942b83b2a00fa285e596e

oxtoacart committed:
Merge commit 'd53b121075374d549aef540ae83d526aef2c6519' into pushupstream

oxtoacart committed:
Merge commit 'edbcd8781f86d8182da52ca5d8789a4129eca175' into pushupstream

oxtoacart committed:
Squashed 'src/github.com/getlantern/chained/' changes from 1be1d6b..fd3c3fa

fd3c3fa Switched connpool, fronted and chained to using interfaces for true immutability of configuration
59b56f0 Updated travis builds to 1.4.1
8a51976 Merge branch 'addlicenses' into peerscanner
303235e Added LICENSES to those projects that needed them
55998d7 Switched to golang.org/x/tools/cmd/cover
cbcc048 Upgraded remaining projects to Go 1.4
ab2ccc8 Making sure to close connIn and connOut in chained Server
3e60d00 Reordered code
7bbf62e Took out unnecessary (and broken) pipelining support

git-subtree-dir: src/github.com/getlantern/chained
git-subtree-split: fd3c3fa3d5500439cbef669511d2ca135ab08e89

oxtoacart committed:
Merge commit '8bfab73407ff3a7cd9fd4cbf78a0cff94d7bb82f' into pushupstream

oxtoacart committed:
Squashed 'src/github.com/getlantern/deepcopy/' changes from 6ca5d81..5b37e86

5b37e86 Updated travis builds to 1.4.1
dc1b55d getlantern/lantern#2083 Switched travis builds to use golang.org/x/tools/cmd/cover
6b04cc7 getlantern/lantern#2083 Switched travis builds to Go 1.4

git-subtree-dir: src/github.com/getlantern/deepcopy
git-subtree-split: 5b37e869d48046c97721cccab5de0ebd12ff94f7

oxtoacart committed:
Merge commit 'd393c72c3dec82eea59c944aa9550deb059af214' into pushupstream

oxtoacart committed:
Squashed 'src/github.com/getlantern/appdir/' changes from 6e5aa6a..6e9e003

6e9e003 getlantern/lantern#2238 Added logging to files with size-based log rotation

git-subtree-dir: src/github.com/getlantern/appdir
git-subtree-split: 6e9e003d3be92239390675718c041acaf7b26399

oxtoacart committed:
Squashed 'src/github.com/getlantern/pathreflect/' changes from 3e7b0e6..5a84f4b

5a84f4b Updated travis builds to 1.4.1
0289f3d getlantern/lantern#2083 Switched travis builds to use golang.org/x/tools/cmd/cover
9c13f16 getlantern/lantern#2083 Switched travis builds to Go 1.4

git-subtree-dir: src/github.com/getlantern/pathreflect
git-subtree-split: 5a84f4bf62efa7b35d5091db87545f669eb7f974

oxtoacart committed:
Merge commit '4e3146543569870142f43d099ad8950f2ede12be' into pushupstream

oxtoacart committed:
Squashed 'src/github.com/getlantern/proxiedsites/' changes from 33aa72f..b4fc115

b4fc115 rename whitelist to proxiedsites everywhere
bfaea41 rename whitelist to proxiedsites everywhere

git-subtree-dir: src/github.com/getlantern/proxiedsites
git-subtree-split: b4fc115b1986a701ec06d84b268228b51aa8fccb

oxtoacart committed:
Merged from proxiedsites upstream

oxtoacart committed:
Merge commit '436dc4c2b16e59bdcd495a368699746979f5abdb' into pushupstream

oxtoacart committed:
Squashed 'src/github.com/getlantern/proxy/' changes from 8e430d7..5c39a61

5c39a61 Switched connpool, fronted and chained to using interfaces for true immutability of configuration
83661f3 Added LICENSES to those projects that needed them

git-subtree-dir: src/github.com/getlantern/proxy
git-subtree-split: 5c39a619c22ec51e7c82f0acd57e511beabf0fd7

oxtoacart committed:
Restored proxiedsites to correct version

oxtoacart committed:
Merge pull request #122 from getlantern/pushupstream

Made changes to support pushing back upstream

oxtoacart committed:
Removed peerscanner_linux_amd64 binary

fffw committed:
add tests

fffw committed:
update test prompt

fffw committed:
update test for iran

fffw committed:
bug fix

fffw committed:
clean up

fffw committed:
bug fix

fffw committed:
add trace

oxtoacart committed:
Made proxy addr in pac file dynamic

oxtoacart committed:
getlantern/lantern#2271 Fixed color of give mode peers

Commit:
Merge pull request #124 from getlantern/issue2271

getlantern/lantern#2271 Fixed color of give mode peers

oxtoacart committed:
Fixed unit test

fffw committed:
refactor code and tests

fffw committed:
add README

fffw committed:
Merge branch 'valencia' into issue2275

fffw committed:
Merge pull request #123 from getlantern/issue2275

Made proxy addr in pac file dynamic

fffw committed:
merge valencia

fffw committed:
update proxiedsites to detour whitelist once lantern.yaml changes

fffw committed:
add detour test to travis

fffw committed:
set detour country to iran

fffw committed:
comment iran detection rules

aranhoide committed:
Merge branch 'valencia' into multifronting-valencia

Conflicts:
	.gitignore
	src/github.com/getlantern/flashlight/config/flags.go
	src/github.com/getlantern/flashlight/flashlight.go
	src/github.com/getlantern/flashlight/server/server.go
	src/github.com/getlantern/peerscanner/cfl/cfl.go
	src/github.com/getlantern/peerscanner/cfl/cfl_test.go
	src/github.com/getlantern/peerscanner/group.go
	src/github.com/getlantern/peerscanner/host.go
	src/github.com/getlantern/peerscanner/main.go

aranhoide committed:
Merge commit '642cb09cc37b91d8593f3df79bd29ed5631ef0bb' as 'src/github.com/getlantern/aws-sdk-go'

aranhoide committed:
Squashed 'src/github.com/getlantern/aws-sdk-go/' content from commit d94a7d4

git-subtree-dir: src/github.com/getlantern/aws-sdk-go
git-subtree-split: d94a7d445cfeef55b129e490a0ff3d67562caab7

aranhoide committed:
Squashed 'src/github.com/vaughan0/go-ini/' content from commit a98ad7e

git-subtree-dir: src/github.com/vaughan0/go-ini
git-subtree-split: a98ad7ee00ec53921f08832bc06ecf7fd600e6a1

aranhoide committed:
Merge commit '7ecff71a54e7a728a253e7dd09f1af6c9b1bb79c' as 'src/github.com/vaughan0/go-ini'

aranhoide committed:
[gost] Added github.com/getlantern/aws-sdk-go/gen/cloudfront and its dependencies

aranhoide committed:
make this compile again after merge carnage

I couldn't test this before because I was missing the AWS library, which
I couldn't `gost get` because I had uncommitted changes.

aranhoide committed:
Squashed 'src/github.com/getlantern/go-dnsimple/' content from commit 6f8bebe

git-subtree-dir: src/github.com/getlantern/go-dnsimple
git-subtree-split: 6f8bebe5ba487c832374d17ff7181cf138894c63

aranhoide committed:
Merge commit 'a1840c14a42cb25bfed8b0651ae62793a9f28dec' as 'src/github.com/getlantern/go-dnsimple'

aranhoide committed:
[gost] Added github.com/getlantern/aws-sdk-go/aws and its dependencies

aranhoide committed:
remove old hardcoded and leaked credentials

aranhoide committed:
minimal unit tests for dsp

myleshorton committed:
Merge pull request #117 from getlantern/issue2284

getlantern/lantern#2284 Hiding hostname from logs

myleshorton committed:
Merge branch 'issue2288' into valencia

myleshorton committed:
Fixing merge conflicts

aranhoide committed:
uniformize credential names

myleshorton committed:
Updated cocoasudo to support OSX 10.6 closes getlantern/lantern#2297

aranhoide committed:
add details to main file comment

aranhoide committed:
make room for dnsimple in main

aranhoide committed:
rename group(s) -> cflGroup(s), to make room for DNSimple groups

aranhoide committed:
remove global hostsByName, rename hostsByIp->hosts

aranhoide committed:
minor renaming

aranhoide committed:
introduce host.dspRecord and some related functionality

aranhoide committed:
fix misunderstanding: `name`s were not fully qualified

aranhoide committed:
add dspgroups

aranhoide committed:
utility to check whether a host is ready to participate in DNSimple roundrobins

aranhoide committed:
finish loadHosts logic for Cloudfront, pending QA

fffw committed:
update copyright in LICENSE

aranhoide committed:
update ListenAndServe in flashlight unit tests

aranhoide committed:
send port 80 to peerdnsreg if running unencrypted

aranhoide committed:
support servers registering with port 80, assume unencrypted

aranhoide committed:
allow Cloudfront when checking reportedHost

aranhoide committed:
Merge branch 'valencia' into multifronting-valencia

Conflicts:
	src/github.com/getlantern/flashlight/README.md

fffw committed:
add comments

fffw committed:
not resend non-idempotent method requests

oxtoacart committed:
getlantern/lantern#2306 Added full list of fallback proxies as chained servers

aranhoide committed:
fixes from QA

aranhoide committed:
update README with new env var names for secrets

aranhoide committed:
Merge pull request #127 from getlantern/issue2306

getlantern/lantern#2306 Added full list of fallback proxies as chained s...

aranhoide committed:
update envvars.bash.enc with new secrets

aranhoide committed:
testing fixes

aranhoide committed:
comments on HostFn and deprecated Host, where they're used

aranhoide committed:
fix comment

aranhoide committed:
resetProxiedClient godoc

aranhoide committed:
friendly error messages when testing without required env vars

aranhoide committed:
tighter scope for dialer

aranhoide committed:
avoid breaking out of register loop on PostForm error

oxtoacart committed:
Merge pull request #130 from getlantern/issue2310

avoid breaking out of register loop on PostForm error

myleshorton committed:
Merge pull request #128 from getlantern/multifronting-valencia

Multifronting valencia

myleshorton committed:
Merge pull request #126 from getlantern/issue2297

Updated cocoasudo to support OSX 10.6 closes getlantern/lantern#2297

fffw committed:
âsend addr to chan when a direct connection closes without error

fffw committed:
update direct access hosts to pac file for speed optimization

fffw committed:
turn off and on to let PAC file take effect

oxtoacart committed:
getlantern/lantern#2296 Refixed after bad merge

myleshorton committed:
Merge pull request #133 from getlantern/issue2296

getlantern/lantern#2296 Refixed after bad merge

fffw committed:
more meaningful name for detector

fffw committed:
reapply pac file only if there's new direct hosts

fffw committed:
adjust logging level in tarfs to avoid confusing

oxtoacart committed:
Merge pull request #134 from getlantern/issue2133

adjust logging level in tarfs to avoid confusing

fffw committed:
use OS fonts to reduce file size

oxtoacart committed:
Merge pull request #114 from getlantern/detour

detect and proxy automatically

oxtoacart committed:
Pointing at new cloud.yaml

myleshorton committed:
Merge pull request #137 from getlantern/fallbackscloud

Pointing at new cloud.yaml that includes chained fallback servers

myleshorton committed:
Merge pull request #135 from getlantern/issue2244

use OS fonts to reduce file size

Commit:
Merge branch 'valencia' into integration/lantern-android-valencia

Commit:
Merge branch 'integration/lantern-android-valencia' into lantern-android

xiam commented:
I'm closing this to target our current development branch (valencia) instead.

xiam commented:
See: https://github.com/getlantern/flashlight-build/pull/139
